### PR TITLE
Respect minimum cores/memory limit during scale down

### PR DESCRIPTION
**Pull request by aleksandra-malinowska**
Follow up to #295, adding support for lower limits of total cluster cores/memory.

### CODE REVIEW

File: cluster-autoscaler/core/scale_down.go [hunk 1]
```diff
@@ -763,3 +806,42 @@ func cleanUpNodeAutoprovisionedGroups(cloudProvider cloudprovider.CloudProvider)
...
+		memoryTotal = memoryTotal + memory
+	}
+
+	return coresTotal, memoryTotal
+}
+
+func filterOutMaster(nodes []*apiv1.Node, pods []*apiv1.Pod) []*apiv1.Node {
+	masters := make(map[string]bool)
+	for _, pod := range pods {
+		if pod.Namespace == "kube-system" && pod.Labels["component"] == "kube-apiserver" {
```
Comment (mwielgus) [inline cluster-autoscaler/core/scale_down.go]: constants please

File: cluster-autoscaler/core/scale_down.go [hunk 2]
```diff
@@ -207,12 +209,18 @@ func (sd *ScaleDown) UpdateUnneededNodes(
 			glog.V(4).Infof("Node %s is not suitable for removal - utilization too big (%f)", node.Name, utilization)
 			continue
 		}
+		cores, memory, err := getNodeCoresAndMemory(node)
+		if err != nil {
+			glog.Warningf("Error getting node resources: %v", err)
+		}
+		coresTotal = coresTotal + cores
```
Comment (mwielgus) [inline cluster-autoscaler/core/scale_down.go]: Move this up.

File: cluster-autoscaler/core/scale_down.go [hunk 3]
```diff
@@ -357,13 +365,16 @@ func (sd *ScaleDown) chooseCandidates(nodes []*apiv1.Node) ([]*apiv1.Node, []*ap
 
 // TryToScaleDown tries to scale down the cluster. It returns ScaleDownResult indicating if any node was
 // removed and error if such occurred.
-func (sd *ScaleDown) TryToScaleDown(nodes []*apiv1.Node, pods []*apiv1.Pod, pdbs []*policyv1.PodDisruptionBudget, currentTime time.Time) (ScaleDownResult, errors.AutoscalerError) {
+func (sd *ScaleDown) TryToScaleDown(allNodes []*apiv1.Node, pods []*apiv1.Pod, pdbs []*policyv1.PodDisruptionBudget, currentTime time.Time) (ScaleDownResult, errors.AutoscalerError) {
 	nodeDeletionDuration := time.Duration(0)
 	findNodesToRemoveDuration := time.Duration(0)
 	defer updateScaleDownMetrics(time.Now(), &findNodesToRemoveDuration, &nodeDeletionDuration)
+	nodes := filterOutMaster(allNodes, pods)
```
Comment (mwielgus) [inline cluster-autoscaler/core/scale_down.go]: nodesWithoutMaster

File: cluster-autoscaler/core/utils.go [hunk 4]
```diff
@@ -415,3 +416,36 @@ func ConfigurePredicateCheckerForLoop(unschedulablePods []*apiv1.Pod, schedulabl
...
 	}
 }
+
+// Getting node cores/memory
+const (
+	// Megabyte is 2^20 bytes.
+	Megabyte float64 = 1024 * 1024
+)
+
+func getNodeCoresAndMemory(node *apiv1.Node) (int64, int64, error) {
```
Comment (mwielgus) [inline cluster-autoscaler/core/utils.go]: please add an explicit test for this

File: cluster-autoscaler/core/utils.go [hunk 5]
```diff
@@ -415,3 +416,36 @@ func ConfigurePredicateCheckerForLoop(unschedulablePods []*apiv1.Pod, schedulabl
...
+
+	if cores <= 0 || memory <= 0 {
+		return 0, 0, fmt.Errorf("Invalid node CPU/memory values - cpu %v, memory %v", cores, memory)
+	}
+
+	memoryMb := math.Ceil(float64(memory) / Megabyte)
+	return cores, int64(memoryMb), nil
+}
+
+func getNodeResource(node *apiv1.Node, resource apiv1.ResourceName) (int64, error) {
```
Comment (mwielgus) [inline cluster-autoscaler/core/utils.go]: please add an explicit test for this

File: cluster-autoscaler/core/scale_down.go [hunk 6]
```diff
@@ -763,3 +806,42 @@ func cleanUpNodeAutoprovisionedGroups(cloudProvider cloudprovider.CloudProvider)
...
+			continue
+		}
+		coresTotal = coresTotal + cores
+		memoryTotal = memoryTotal + memory
+	}
+
+	return coresTotal, memoryTotal
+}
+
+func filterOutMaster(nodes []*apiv1.Node, pods []*apiv1.Pod) []*apiv1.Node {
```
Comment (mwielgus) [inline cluster-autoscaler/core/scale_down.go]: explicit test please

File: cluster-autoscaler/core/scale_down.go [hunk 7]
```diff
@@ -763,3 +806,42 @@ func cleanUpNodeAutoprovisionedGroups(cloudProvider cloudprovider.CloudProvider)
 	}
 	return nil
 }
+
+func (sd *ScaleDown) calculateCoresAndMemoryTotal(nodes []*apiv1.Node, timestamp time.Time) (int64, int64) {
```
Comment (mwielgus) [inline cluster-autoscaler/core/scale_down.go]: explicit test please

File: cluster-autoscaler/core/utils.go (hunk 4 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/core/utils.go]: done

File: cluster-autoscaler/core/utils.go (hunk 5 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/core/utils.go]: done

File: cluster-autoscaler/core/scale_down.go (hunk 1 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/core/scale_down.go]: done

File: cluster-autoscaler/core/scale_down.go (hunk 2 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/core/scale_down.go]: Replaced with constant as discussed offline

File: cluster-autoscaler/core/scale_down.go (hunk 3 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/core/scale_down.go]: done

File: cluster-autoscaler/core/scale_down.go (hunk 6 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/core/scale_down.go]: done

File: cluster-autoscaler/core/scale_down.go (hunk 7 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/core/scale_down.go]: done

**Comment by mwielgus** [comment]
```
# k8s.io/autoscaler/cluster-autoscaler/core
core/scale_down.go:217: undefined: config.DefaultMaxClusterCores
core/scale_down.go:217: undefined: config.DefaultMaxClusterMemory
godep: go exit status 2
```

**Comment by aleksandra-malinowska** [comment]
@mwielgus forgot to commit a new file, sorry:) should be fixed now

**Pull request merged**